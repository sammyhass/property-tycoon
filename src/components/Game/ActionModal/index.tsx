import { Box, Heading } from '@chakra-ui/react';
import React, { useMemo } from 'react';
import HUD from '../HUD';
import {
  ActionModalBuy,
  ActionModalFreeParking,
  ActionModalGetOutJail,
  ActionModalGo,
  ActionModalGoToJail,
  ActionModalPayRent,
  ActionModalRoll,
  ActionModalTakeCard,
  ActionModalTax,
} from './Content';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'GET_OUT_OF_JAIL' // Player attempts to get out of jail (shows instead)
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'TAX' // Pay tax
  // TODO:
  | 'MORTGAGE'
  | 'GO'; // Land on a Go space

export interface ActionModalProps {
  action: ActionType | null;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: ActionModalBuy,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: ActionModalPayRent,
  GET_OUT_OF_JAIL: ActionModalGetOutJail,
  ROLL: ActionModalRoll,
  GO_TO_JAIL: ActionModalGoToJail,
  TAKE_POT_LUCK: () => <ActionModalTakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: ActionModalFreeParking,
  GO: ActionModalGo,
  TAX: ActionModalTax,
};

export default function ActionSidebar(props: ActionModalProps) {
  const ActionComponent = useMemo(
    () =>
      props.action
        ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
        : null,
    [props.action]
  );

  return (
    <Box minW={'400px'} bg="white" px="20px" h="100vh">
      <Heading>
        {props.action
          ?.split('_')
          .map(w => w[0].toUpperCase() + w.slice(1).toLowerCase())
          .join(' ')}
      </Heading>
      <Box minH="70%">{ActionComponent && <ActionComponent {...props} />}</Box>
      <HUD />
    </Box>
  );
}
