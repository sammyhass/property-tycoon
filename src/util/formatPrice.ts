// Price formatting to use throughout the app.
export const formatPrice = (value: number) => `Â£${value}`;
