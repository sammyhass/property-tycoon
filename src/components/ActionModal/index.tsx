import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Box,
  Button,
  Divider,
  Flex,
  Heading,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useState } from 'react';
import BoardSpace from '../Board/spaces';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'MORTGAGE'
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'GO'; // Land on a Go space

interface ActionModalProps {
  action: ActionType | null;
  onClose: () => void;
  isOpen: boolean;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: () => <ActionModal.Buy />,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: () => <div>Pay Rent</div>,
  ROLL: props => <ActionModal.Roll {...props} />,
  GO_TO_JAIL: () => <div>Jail</div>,
  TAKE_POT_LUCK: () => <ActionModal.TakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModal.TakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: () => <div>Free Parking</div>,
  GO: () => <div>GO!</div>,
};

export default function ActionModal(props: ActionModalProps) {
  const ActionComponent = props.action
    ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
    : null;

  return (
    <Modal onClose={props.onClose} isOpen={props.isOpen}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>{props.action}</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          {ActionComponent && <ActionComponent {...props} />}
        </ModalBody>
      </ModalContent>
    </Modal>
  );
}

// Content for our action modal when a player is rolling the dice.
ActionModal.Roll = ({ onClose }: ActionModalProps) => {
  const { currentPlayer, move } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [roll, setRoll] = useState(0);

  // Roll the dice.
  const handleRoll = () => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 6
    const roll = Math.floor(Math.random() * 6) + 1;

    setTimeout(() => {
      move(currentPlayer?.token, roll);
      setRoll(roll);
      setIsRolling(false);
    }, 1000);

    if (!currentPlayer) {
      return <></>;
    }
  };
  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {roll > 0 && <Heading size="sm">Dice landed on {roll}</Heading>}
      <Button onClick={() => handleRoll()}>Roll</Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
ActionModal.TakeCard = ({ cardType }: { cardType: CardType }) => {
  const { takeCard } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      <pre>{JSON.stringify(cardAction, null, 2)}</pre>
      <Button onClick={() => handleTakeCard()}>Take</Button>
    </Flex>
  );
};

ActionModal.Buy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <Stack>
          <Text p="0" fontWeight={'600'}>
            Rent Unimproved: {formatPrice(property.rent_unimproved ?? 0)}
            <br />
            Rent 1 House: {formatPrice(property.rent_one_house ?? 0)}
            <br />
            Rent 2 Houses: {formatPrice(property.rent_two_house ?? 0)}
            <br />
            Rent 3 Houses: {formatPrice(property.rent_three_house ?? 0)}
            <br />
            Rent 4 Houses: {formatPrice(property.rent_four_house ?? 0)}
            <br />
          </Text>
        </Stack>
      </Box>
      <Button
        colorScheme="green"
        isLoading={buying}
        onClick={() => {
          handleBuy();
        }}
      >
        Buy {property.name}
      </Button>
    </Box>
  );
};
