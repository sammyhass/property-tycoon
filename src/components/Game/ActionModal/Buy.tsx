import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { Box, Button, Divider, Flex, Heading } from '@chakra-ui/react';
import React, { useState } from 'react';
import { PropertyRentInfo } from '../Board/PropertyStack';
import { BoardSpaceProperty } from '../Board/spaces';

export default function BuyContent() {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.
  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  const group = gameSettings?.PropertyGroups.find(
    group => group.color === property?.property_group_color
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpaceProperty property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <PropertyRentInfo property={property} group={group} />
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
}
