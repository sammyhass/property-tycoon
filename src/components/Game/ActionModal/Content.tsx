import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Alert,
  AlertDescription,
  Box,
  Button,
  Divider,
  Flex,
  FormControl,
  FormLabel,
  Heading,
  Radio,
  RadioGroup,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { ActionModalProps } from '.';
import GameCard from '../Board/cards/Card';
import { PropertyRentInfo } from '../Board/PropertyStack';
import BoardSpace from '../Board/spaces';

// Action Modal Content Components

export const ActionModalGetOutJail = ({ action }: ActionModalProps) => {
  const {
    currentPlayer,
    rollDice,
    payBank,
    state,
    hideActionModal,
    move,
    freeFromJail,
    failedToGetOutOfJail,
  } = useGameContext();

  const [choice, setChoice] = useState<'roll' | 'pay'>('roll');

  const [roll, setRoll] = useState<[number, number] | null>(null);

  const [didRoll, setDidRoll] = useState(false);
  const [didPay, setDidPay] = useState(false);

  const handleContinue = useCallback(() => {
    if (!currentPlayer) return;
    if (didPay || (roll && roll[0] === roll[1])) {
      freeFromJail(currentPlayer.token);
      if (roll && roll[0] === roll[1]) {
        move(currentPlayer.token, roll[0] + roll[1]);
      }
    } else {
      failedToGetOutOfJail(currentPlayer.token);
    }
    hideActionModal();
  }, [
    failedToGetOutOfJail,
    currentPlayer,
    hideActionModal,
    move,
    roll,
    didPay,
    freeFromJail,
  ]);

  const handleDiceRoll = useCallback(() => {
    if (!currentPlayer) return;
    const [roll1, roll2] = rollDice();
    setRoll([roll1, roll2]);
    setDidRoll(true);
  }, [rollDice, currentPlayer]);

  const handlePayToGetOut = useCallback(() => {
    if (!currentPlayer) return;
    payBank(currentPlayer?.token, 50);
    setDidPay(true);
  }, [currentPlayer, payBank]);

  if (!currentPlayer) return <></>;
  return (
    <Flex direction={'column'} gap="5px">
      <Heading size="md" py="10px">
        You&apos;sre in Jail!
      </Heading>

      {!(didRoll || didPay) && (
        <FormControl mb="20px">
          <FormLabel>
            <Text fontSize="sm">
              Try to get out of jail by rolling doubles (
              {3 - (state[currentPlayer.token]?.turnsInJail ?? 0)} tries left)
              or paying the fine of $50.
            </Text>
          </FormLabel>
          <RadioGroup
            value={choice}
            onChange={v =>
              v === 'roll' ? setChoice(v) : v === 'pay' ? setChoice(v) : null
            }
            defaultValue="roll"
          >
            <Stack direction="column" spacing="5px">
              <Radio value="roll">
                <Text>ðŸŽ² Roll Dice</Text>
              </Radio>
              <Radio value="pay">
                <Text>ðŸ’¸ Pay $50 to get out</Text>
              </Radio>
            </Stack>
          </RadioGroup>
        </FormControl>
      )}

      {choice === 'roll' ? (
        <Box>
          <Heading size="sm">
            Try to get out of jail by rolling doubles (
            {3 - (state[currentPlayer.token]?.turnsInJail ?? 0)} tries left)
          </Heading>
          <Box p="10px" bg="#eee" borderRadius={'8px'}>
            <Heading textAlign={'center'}>ðŸŽ² ðŸŽ²</Heading>
            {roll ? (
              <Box>
                <Text textAlign={'center'}>You Rolled</Text>
                <Stack
                  direction="row"
                  spacing={'5px'}
                  justify="center"
                  fontSize={'lg'}
                >
                  <Text fontWeight={'bold'}>{roll[0]}</Text>
                  <Text>and</Text>
                  <Text fontWeight={'bold'}>{roll[1]}</Text>
                </Stack>
                {roll[0] === roll[1] && (
                  <Alert status="success">
                    <AlertDescription>
                      You rolled doubles! You can get out of jail!
                    </AlertDescription>
                  </Alert>
                )}
              </Box>
            ) : (
              <Button onClick={handleDiceRoll} w="100%" colorScheme={'red'}>
                Roll Dice
              </Button>
            )}
          </Box>
        </Box>
      ) : (
        <Box>
          <Heading size="sm">Pay Â£50 to get out of jail now?</Heading>
          <Button
            colorScheme={'purple'}
            onClick={handlePayToGetOut}
            w="100%"
            isDisabled={
              !currentPlayer || (state[currentPlayer.token]?.money ?? 0) < 50
            }
          >
            Pay Â£50 to get out
          </Button>
        </Box>
      )}
      {(didRoll || didPay) && (
        <Button onClick={handleContinue}>
          {roll
            ? roll[0] === roll[1] || didPay
              ? 'Get out Now'
              : 'Roll again next turn.'
            : 'Ok'}
        </Button>
      )}
    </Flex>
  );
};

// Content for our action modal when a player is rolling the dice.
export const ActionModalRoll = () => {
  const { currentPlayer, move, rollDice } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [isMoving, setIsMoving] = useState(false);

  const [diceRoll, setDiceRoll] = useState<[number, number] | null>(null);

  const handleMove = useCallback(() => {
    if (!currentPlayer) return;
    setIsMoving(true);
    setTimeout(() => {
      move(currentPlayer.token, diceRoll![0] + diceRoll![1]);
      setIsMoving(false);
    }, 500);
  }, [currentPlayer, diceRoll, move]);

  // Roll the dice.
  const handleRoll = useCallback(() => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 12
    setTimeout(() => {
      const [dice1, dice2] = rollDice();
      setDiceRoll([dice1, dice2]);
      setIsRolling(false);
    }, 500);
  }, [currentPlayer, rollDice]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ² ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {diceRoll && (
        <Stack spacing={4} justify="center" align={'center'}>
          <Text>You rolled:</Text>
          <Flex justify={'center'} gap="10px" align="center">
            <Heading size="md">{diceRoll[0]}</Heading>
            <Text>{` and `}</Text>
            <Heading size="md">{diceRoll[1]}</Heading>
          </Flex>
        </Stack>
      )}
      <Button
        w="100%"
        colorScheme={diceRoll ? 'green' : 'purple'}
        onClick={diceRoll ? handleMove : handleRoll}
        mt="10px"
        isLoading={isRolling || isMoving}
        disabled={isRolling || isMoving}
      >
        {diceRoll ? 'Move' : 'Roll'}
      </Button>
    </Flex>
  );
};

export const ActionModalGo = ({}) => {
  const { payBank, currentPlayer, hideActionModal } = useGameContext();

  const [loading, setLoading] = useState(false);

  const collect = useCallback(() => {
    setLoading(true);
    if (!currentPlayer) return;
    payBank(currentPlayer.token, -200);

    setTimeout(() => {
      setLoading(false);

      hideActionModal();
    }, 500);
  }, [currentPlayer, payBank, hideActionModal]);

  if (!currentPlayer) return <></>;

  return (
    <Flex direction={'column'} gap="5px" align={'center'}>
      <Heading size="4xl">ðŸŽ‰</Heading>
      <Heading size="md">You landed on Go!</Heading>
      <Box>Collect {formatPrice(200)} from the bank!</Box>
      <Button
        w="100%"
        colorScheme={'green'}
        isLoading={loading}
        onClick={collect}
      >
        Collect Â£200
      </Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
export const ActionModalTakeCard = ({ cardType }: { cardType: CardType }) => {
  const {
    takeCard,
    payBank,
    currentPlayer,
    gameSettings,
    goto,
    sendToJail,
    hideActionModal,
  } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  const performAction = useCallback(() => {
    if (!cardAction || !currentPlayer?.token) return;

    hideActionModal();

    switch (cardAction.action_type) {
      case 'EARN_FROM_BANK':
        payBank(currentPlayer?.token, -(cardAction?.cost ?? 0));
        break;
      case 'PAY_BANK':
        payBank(currentPlayer?.token, cardAction?.cost ?? 0);
        break;
      case 'GO_TO_JAIL':
        sendToJail(currentPlayer?.token);
        break;
      case 'GO_TO_PROPERTY':
        const property = gameSettings?.BoardSpaces?.find(
          space => space.property_id === cardAction?.property_id
        );
        if (property) {
          goto(currentPlayer?.token, property.board_position);
        }
      default:
        break;
    }
  }, [
    cardAction,
    currentPlayer,
    gameSettings,
    hideActionModal,
    payBank,
    sendToJail,
    goto,
  ]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      {cardAction && (
        <GameCard
          {...cardAction}
          propertyName={
            gameSettings?.Properties?.find(
              property => property.id === cardAction?.property_id
            )?.name ?? undefined
          }
        />
      )}
      <Button
        colorScheme={'green'}
        mt={'10px'}
        w="100%"
        onClick={() => (cardAction ? performAction() : handleTakeCard())}
      >
        {cardAction ? 'Perform Card Action' : 'Take'}
      </Button>
    </Flex>
  );
};

export const ActionModalBuy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <PropertyRentInfo property={property} />
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
};

export const ActionModalTax = () => {
  const { gameSettings, state, currentPlayer, payBank, hideActionModal } =
    useGameContext();

  if (!gameSettings || !currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
  );

  const taxAmount = space?.tax_cost ?? 0;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Tax</Heading>
      <Text fontSize={'4xl'}>{formatPrice(taxAmount)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          payBank(currentPlayer?.token, taxAmount);
          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalPayRent = () => {
  const {
    payPlayer,
    gameSettings,
    currentPlayer,
    state,
    isOwned,
    calculateRent,
    hideActionModal,
  } = useGameContext();

  const space = currentPlayer?.token
    ? gameSettings?.BoardSpaces.find(
        space =>
          space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
      )
    : null;

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  const owner = isOwned(property?.id ?? '');

  const rent = useMemo(() => {
    if (!property) return 0;
    return calculateRent(property?.id);
  }, [calculateRent, property]);

  if (!owner || !property) return <></>;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Rent</Heading>
      {property && <BoardSpace.Property property={property} />}
      <Text fontSize={'4xl'}>{formatPrice(rent ?? 0)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          if (!currentPlayer) return;
          payPlayer(currentPlayer?.token, owner?.token, rent);

          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalFreeParking = () => {
  const {
    landedOnFreeParking,
    currentPlayer,
    hideActionModal,
    totalOnFreeParking,
  } = useGameContext();

  useEffect(() => {
    if (!currentPlayer) return;
    const v = landedOnFreeParking(currentPlayer.token);
  }, [currentPlayer, landedOnFreeParking]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Free Parking</Heading>

      <Text>
        The total amount on free parking is {formatPrice(totalOnFreeParking)},
        It&apos;s all yours now
      </Text>
      <Button
        w={'100%'}
        colorScheme={'blue'}
        onClick={() => {
          hideActionModal();
        }}
      >
        Continue
      </Button>
    </Flex>
  );
};

export const ActionModalGoToJail = () => {
  const { sendToJail, currentPlayer, hideActionModal, gameSettings } =
    useGameContext();

  const handleGoToJail = () => {
    if (!currentPlayer) return;
    sendToJail(currentPlayer?.token);

    setTimeout(() => {
      hideActionModal();
    }, 1000);
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Go To Jail</Heading>
      <Text>You have been sent to jail</Text>
      <Button onClick={() => handleGoToJail()}>Go To Jail</Button>
    </Flex>
  );
};
