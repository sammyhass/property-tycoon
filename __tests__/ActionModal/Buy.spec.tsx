import { ActionModalBuy } from '@/components/Game/ActionModal/Content';
import { PlayerState, TokenType } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { act } from '@testing-library/react';
import React from 'react';
import { fakeGameSettings, fakePlayer, renderWithGameContext } from '../fakers';

describe('Buy ActionModal', () => {
  it('matches snapshot', () => {
    const { container } = renderWithGameContext(
      <ActionModalBuy />,

      {
        currentPlayer: fakePlayer('cat'),
        gameSettings: fakeGameSettings(),
        state: {
          boot: {
            inJail: false,
            money: 400,
            pos: 1,
            propertiesOwned: [],
            lastRoll: 4,
            turnsInJail: 0,
          },
          cat: {
            inJail: false,
            money: 1200,
            lastRoll: 6,
            turnsInJail: 0,
            pos: 1,
            propertiesOwned: [],
          },
        },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correct property information', () => {
    const gameSettings = fakeGameSettings();
    const { container } = renderWithGameContext(<ActionModalBuy />, {
      currentPlayer: fakePlayer('cat'),
      gameSettings,
      state: {
        boot: {
          inJail: false,
          money: 400,
          lastRoll: 5,
          turnsInJail: 0,
          pos: 1,
          propertiesOwned: [],
        },
        cat: {
          inJail: false,
          lastRoll: 11,
          turnsInJail: 0,
          money: 1200,
          pos: 1,
          propertiesOwned: [],
        },
      },
    });
    expect(container.textContent?.toUpperCase()).toContain('BUY');

    expect(container.textContent?.toUpperCase()).toContain(
      gameSettings?.Properties?.[0]?.name?.toUpperCase()
    );
    expect(container.textContent?.toUpperCase()).toContain(
      formatPrice(gameSettings?.Properties?.[0]?.price ?? 0)
    );
  });

  it('current player can buy property and property gets added to their owned properties', async () => {
    const gameSettings = fakeGameSettings();

    // reactive state
    const state: PlayerState = {
      boot: {
        inJail: false,
        lastRoll: 5,
        turnsInJail: 0,
        money: 400,
        pos: 1,
        propertiesOwned: [],
      },
      cat: {
        inJail: false,
        money: 1200,
        pos: 1,
        propertiesOwned: [],
        lastRoll: 11,
        turnsInJail: 0,
      },
    };

    const buy = jest
      .fn()
      .mockImplementation(async (player: TokenType, propertyId) => {
        state[player]?.propertiesOwned.push(propertyId);
      });

    const { container } = renderWithGameContext(<ActionModalBuy />, {
      currentPlayer: fakePlayer('cat'),
      gameSettings,
      buy,
      state,
    });

    const buyButton = container.querySelectorAll('button')[0];
    await act(async () => {
      buyButton.click();
      await new Promise(resolve => setTimeout(resolve, 1200));
    });

    expect(buy).toHaveBeenCalledWith('cat', gameSettings?.Properties?.[0].id);
    expect(state.cat?.propertiesOwned).toContain(
      gameSettings?.Properties?.[0].id
    );
  });
});
