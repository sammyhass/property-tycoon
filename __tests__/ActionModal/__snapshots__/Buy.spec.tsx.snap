// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buy ActionModal matches snapshot 1`] = `
<div
  class="css-0"
>
  <div
    class="css-u6v8er"
  >
    <div
      class="boardSpace property css-0"
      style="background: rgb(0, 0, 255);"
    >
      <div
        class="boardBackground"
      />
      <div
        class="propertyContent"
      >
        <div
          class="css-1lads1q"
        />
        <div
          class="propertyText"
        >
          Property 1
        </div>
        <div
          class="propertyPrice"
        >
          ¬£100
        </div>
      </div>
    </div>
  </div>
  <hr
    aria-orientation="horizontal"
    class="chakra-divider css-1g0y1ze"
  />
  <div
    class="css-ey3b4m"
  >
    <h2
      class="chakra-heading css-0"
    >
      Property 1
       - 
      ¬£100
    </h2>
    <div
      class="css-21wlw6"
    >
      <h2
        class="chakra-heading css-0"
      >
        Rent
      </h2>
      <ul
        class="css-155za0w"
        role="list"
      >
        <li
          class="css-0"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-money-bill fa-w-20 "
              color="#00f"
              data-icon="money-bill"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M608 64H32C14.33 64 0 78.33 0 96v320c0 17.67 14.33 32 32 32h576c17.67 0 32-14.33 32-32V96c0-17.67-14.33-32-32-32zM48 400v-64c35.35 0 64 28.65 64 64H48zm0-224v-64h64c0 35.35-28.65 64-64 64zm272 176c-44.19 0-80-42.99-80-96 0-53.02 35.82-96 80-96s80 42.98 80 96c0 53.03-35.83 96-80 96zm272 48h-64c0-35.35 28.65-64 64-64v64zm0-224c-35.35 0-64-28.65-64-64h64v64z"
                fill="currentColor"
              />
            </svg>
          </svg>
          When 
          1
           owned, rent is 
          4
          x Dice Roll.
        </li>
        <li
          class="css-0"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-money-bill fa-w-20 "
              color="#00f"
              data-icon="money-bill"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M608 64H32C14.33 64 0 78.33 0 96v320c0 17.67 14.33 32 32 32h576c17.67 0 32-14.33 32-32V96c0-17.67-14.33-32-32-32zM48 400v-64c35.35 0 64 28.65 64 64H48zm0-224v-64h64c0 35.35-28.65 64-64 64zm272 176c-44.19 0-80-42.99-80-96 0-53.02 35.82-96 80-96s80 42.98 80 96c0 53.03-35.83 96-80 96zm272 48h-64c0-35.35 28.65-64 64-64v64zm0-224c-35.35 0-64-28.65-64-64h64v64z"
                fill="currentColor"
              />
            </svg>
          </svg>
          When 
          2
           owned, rent is 
          10
          x Dice Roll.
        </li>
      </ul>
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-1y9z5uc"
      />
      <div
        class="css-0"
      >
        <h2
          class="chakra-heading css-0"
        >
          Houses and Hotels
        </h2>
        <p
          class="chakra-text css-0"
        >
          üè† House Cost: 
          ¬£100
        </p>
        <p
          class="chakra-text css-0"
        >
          üè® Hotel Cost: 
          ¬£100
        </p>
      </div>
    </div>
  </div>
  <div
    class="css-1gnc3qn"
  >
    <button
      class="chakra-button css-uxt1e8"
      type="button"
    >
      Buy
    </button>
    <button
      class="chakra-button css-uxt1e8"
      type="button"
    >
      No thanks
    </button>
  </div>
</div>
`;
